Database Redo

xales opened the pull request:
This is the long-awaited complete rework of the database. **Note: not complete yet**

Notably:

- This is the first step in the config being axed.
- All "legacy" systems for admins, bans, whitelist, etc. are being removed, and a database is now required.
- If you don't configure an external database (MySQL is no longer supported, PostgreSQL is instead), a SQLite database will be used instead.
- Localhost and seeker key will automatically be made admin to ease configuration of database.
- Script will be provided for easy maintenance of sqlite database (essentially, add/remove admin only, anything else will be done in game).
- The sqlite database implementation is intentionally limited; if your server is big enough for e.g. configured job restrictions by player age, you should set up a real database and not rely on sqlite (intended for small server or local testing).
- Bans are being reworked entirely. Bans are now checked by calling `is_banned(something)` on the client. The `something` can be a list, or not, and all used strings aside from direct job/antag titles are being collected as defines. Working on further improvements to this as I think of them.
- Jobbans, whitelist, and admin status are reloaded upon reconnect.
- The Statistics system has been removed, mostly in a single commit. Baystation has never used this and I don't like how it's implemented (RDBMS is not great for time series data).
- The ban and admin permissions tools of old have been completely removed. I am now working on rewriting them to be better.
- The syndicate beacon (calls singularity to it) was removed because I happened across it and it doesn't really fit/exist on our server. More of these purges to come later, but I was eager to get started.
- The admin newscaster was removed because I'm vindictive and the code was bad and polluting admin files. I'll put it back in admin rework thingy.

Most importantly:

**The followup to this commit will store characters in the database as JSON objects. Following the merge and migration (likely with a test in there), character saves will no longer be lost randomly.**

xales committed:
WIP

xales committed:
Squashed 'code/modules/dmpg/' content from commit dc0a3d93ad

git-subtree-dir: code/modules/dmpg
git-subtree-split: dc0a3d93ad4d0b075e4af438b72326a40434284f

xales committed:
Merge commit 'ac9849e284fcc8dac326a741c1c4b1c649d60ac8' as 'code/modules/dmpg'

xales committed:
WIP

xales committed:
WIP Remove Statistics

xales committed:
WIP: compiles

xales committed:
WIP: next

xales committed:
Merge

xales committed:
Holder redo

xales committed:
Fix fax, ban check

xales committed:
Add ban scope recording

xales committed:
More banning scopes

xales committed:
Rank + branch bans

xales commented:
Incidentally, this now supports bans by department, military branch, and military rank.

xales committed:
Squashed 'code/modules/dmpg/' changes from dc0a3d93ad..60aeb9e859

60aeb9e859 Use tabs

git-subtree-dir: code/modules/dmpg
git-subtree-split: 60aeb9e859a66434bd3edc03e93d4cdddee7b998

xales committed:
Merge commit '1d5e381315310fd9e5027b36d12adff8b29e7151' into db_redo

xales committed:
Better satisfy code quality checks

xales committed:
Update globals, example config

Lohikar commented:
Already said all this on discord, but so it's not forgotten/missed:
- The newline changes should be split into their own PR if done at all
- The rename of `/datum/admins` should be split into its own PR
- The deletion of SSstatistics should be split into its own PR
- The rust code should be moved into the `lib` dir
- The rust code should compile on stable rust (no feature flags)
- The deletion of the admin newscaster and the syndie beacon should be documented

xales committed:
Move dmpg to lib dir

Squashed 'lib/dmpg/' content from commit 881d075a7a

git-subtree-dir: lib/dmpg
git-subtree-split: 881d075a7acba37ee304ee8f1f043020fabdcf07

xales commented:
Replaced by #25174 
Duplicate of #25174
